File: disco.py
from flask import Flask, Response
import requests
import json
import urllib3

app = Flask(__name__)

# -- PANEL BİLGİLERİ --
HOST = "https://ornek.paneladresi.com"
USERNAME = "KULLANICI_ADINIZ"
PASSWORD = "SIFRENIZ"

# Güvenlik / istek ayarları
VERIFY_SSL = True  # Eğer self-signed sertifika kullanıyorsan False yapabilirsin, ancak güvenlik riski vardır.
REQUEST_TIMEOUT = 10  # saniye

if not VERIFY_SSL:
    # verify=False kullanıyorsan InsecureRequestWarning'ı kapat
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

API_BASE = f"{HOST}/player_api.php"

# -- HTTP BAŞLIKLARI --
API_HEADERS = {
    'User-Agent': 'Dalvik/2.1.0 (Linux; U; Android 9; SmartBox Build/PI)',
    'Accept': 'application/json',
    'Referer': f'{HOST}/',
}


def _safe_json_resp(r):
    """HTTP durumunu kontrol et ve JSON parse et; hata mesajı döndür."""
    if r.status_code != 200:
        return None, f"HTTP error: {r.status_code}"
    try:
        return r.json(), None
    except json.JSONDecodeError:
        return None, "JSON decode error"


def get_streams():
    """Tüm canlı akışları ve kategorileri çek.
    Döner: (categories_list, streams_list, error_message_or_None)
    """
    # 1. Kategorileri Çekme
    params_cat = {
        'username': USERNAME,
        'password': PASSWORD,
        'action': 'get_live_categories'
    }
    try:
        r_cat = requests.get(API_BASE, params=params_cat, headers=API_HEADERS, timeout=REQUEST_TIMEOUT, verify=VERIFY_SSL)
    except requests.RequestException as e:
        return None, None, f"Kategori isteği başarısız: {e}"

    categories, err = _safe_json_resp(r_cat)
    if err:
        return None, None, f"Kategori hatası: {err}"

    # 2. Kanalları Çekme
    params_stream = {
        'username': USERNAME,
        'password': PASSWORD,
        'action': 'get_live_streams'
    }
    try:
        r_stream = requests.get(API_BASE, params=params_stream, headers=API_HEADERS, timeout=REQUEST_TIMEOUT, verify=VERIFY_SSL)
    except requests.RequestException as e:
        return None, None, f"Kanal isteği başarısız: {e}"

    streams, err = _safe_json_resp(r_stream)
    if err:
        return None, None, f"Kanal hatası: {err}"

    # Beklenen format liste olduğu için kontrol (bazı paneller dict içinde 'categories' vb dönebilir)
    if isinstance(categories, dict) and 'categories' in categories:
        categories = categories['categories']
    if isinstance(streams, dict) and 'live_streams' in streams:
        streams = streams['live_streams']

    if not isinstance(categories, list) or not isinstance(streams, list):
        return None, None, "Beklenmeyen API formatı"

    return categories, streams, None


@app.route('/m3u-dynamic')
def generate_m3u():
    categories, streams, error = get_streams()

    if error:
        return Response(error, mimetype='text/plain', status=500)

    # Kategori ID'den isme eşleştirme (türleri normalize et)
    category_map = {}
    for c in categories:
        cid = str(c.get('category_id', ''))
        cname = c.get('category_name', 'Diğer')
        category_map[cid] = cname

    m3u_lines = ["#EXTM3U"]

    for stream in streams:
        stream_id = stream.get('stream_id')
        name = stream.get('name') or f"channel_{stream_id}"
        icon = stream.get('stream_icon', '') or ''
        category_id = str(stream.get('category_id', ''))
        category_name = category_map.get(category_id, 'Diğer')
        container = stream.get('container_extension') or 'ts'

        # Basit kaçış: double-quote'ları bozmayalım
        name_esc = name.replace('"', '\\"')
        icon_esc = icon.replace('"', '\\"')

        stream_url = f"{HOST}/live/{USERNAME}/{PASSWORD}/{stream_id}.{container}"

        m3u_lines.append(f'#EXTINF:-1 tvg-id="{stream_id}" tvg-name="{name_esc}" tvg-logo="{icon_esc}" group-title="{category_name}",{name_esc}')
        m3u_lines.append(stream_url)

    m3u_content = "\n".join(m3u_lines) + "\n"

    return Response(
        m3u_content,
        mimetype='application/x-mpegurl; charset=utf-8',
        headers={"Content-Disposition": "attachment; filename=playlist.m3u"}
    )


if __name__ == '__main__':
    # Production'da gunicorn/uWSGI kullan; development için:
    app.run(host='0.0.0.0', port=5000, debug=False)

File: disco.php
<?php
// M3U ÇIKTISI BAŞLIĞI
// Tarayıcıya veya oynatıcıya bunun bir M3U dosyası olduğunu bildirir.
header('Content-Type: application/octet-stream');
header('Content-Disposition: attachment; filename="playlist.m3u"');

// -----------------------------------------------------------
// 1. KULLANICI BİLGİLERİNİZİ BURAYA GİRİN
// -----------------------------------------------------------

// Lütfen bu kısımları kendi bilgilerinizle değiştirin.
$host = "https://goldvod.org"; // Başında http:// veya https:// olmalı
$username = "hpgdisco";
$password = "123456";

// API uç noktası ve kimlik doğrulama linkleri
$api_link = $host . "/player_api.php?username=" . $username . "&password=" . $password;
$live_categories_link = $api_link . "&action=get_live_categories";
$live_streams_link = $api_link . "&action=get_live_streams";

// M3U dosyasının başlangıcı
echo "#EXTM3U\n";

// -----------------------------------------------------------
// 2. HTTP İSTEĞİ (cURL) FONKSİYONU
// -----------------------------------------------------------
// PHP'de cURL, en güvenilir ve terminaldeki cURL komutlarına en yakın istek yöntemidir.

function curl_request($url, $custom_headers) {
    // cURL başlat
    $ch = curl_init();
    
    // URL ve Temel Ayarlar
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1); // Yanıtı string olarak döndür
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false); // SSL doğrulamayı yoksay (Gerekiyorsa, terminaldeki -k gibi)
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true); // Yönlendirmeleri takip et
    curl_setopt($ch, CURLOPT_TIMEOUT, 15); // Zaman aşımı (saniye)

    // Başlıkları Ekleme (Mobil Uygulama Taklidi)
    curl_setopt($ch, CURLOPT_HTTPHEADER, $custom_headers);

    // İsteği Çalıştır
    $response = curl_exec($ch);
    
    // Hata kontrolü
    if (curl_errno($ch)) {
        return "HATA: cURL hatası - " . curl_error($ch);
    }
    
    // cURL kapat
    curl_close($ch);
    
    return $response;
}

// Başlıklar: Terminal/Mobil uygulama taklidi (Önceki konuşmalarımızdaki en iyi sonuç veren başlıklar)
$api_headers = array(
    'User-Agent: Dalvik/2.1.0 (Linux; U; Android 9; SmartBox Build/PI)',
    'Accept: application/json',
    'Referer: ' . $host . '/',
);

// -----------------------------------------------------------
// 3. API VERİLERİNİ ÇEKME VE İŞLEME
// -----------------------------------------------------------

// A. Kategori Listesini Çek
$categories_json = curl_request($live_categories_link, $api_headers);

// B. JSON Temizleme ve Ayrıştırma (Önceki sorun giderme adımı)
$categories_json = trim($categories_json);
$categories_json = preg_replace('/^[\x{FEFF}]/', '', $categories_json); // BOM temizleme
$categories_json = preg_replace('/^.*?\{/', '{', $categories_json); // JSON başlamadan önceki çöpleri silme

$categories = json_decode($categories_json, true);

if (json_last_error() !== JSON_ERROR_NONE || !is_array($categories)) {
    // Hata oluşursa M3U dosyasına hata mesajı ekleyip durdur
    echo "# HATA: API Kategori verisi çekilemedi veya JSON bozuk: " . json_last_error_msg() . "\n";
    exit;
}

// C. Tüm Kanalları Çek
$streams_json = curl_request($live_streams_link, $api_headers);
$streams_json = trim($streams_json);
$streams_json = preg_replace('/^[\x{FEFF}]/', '', $streams_json); 
$streams_json = preg_replace('/^.*?\{/', '{', $streams_json);

$streams = json_decode($streams_json, true);

if (json_last_error() !== JSON_ERROR_NONE || !is_array($streams)) {
    echo "# HATA: API Kanal verisi çekilemedi veya JSON bozuk: " . json_last_error_msg() . "\n";
    exit;
}

// Kategori ID'leri için hızlı arama dizisi oluştur (stream_id => category_id)
$category_map = [];
foreach ($categories as $category) {
    $category_map[$category['category_id']] = $category['category_name'];
}


// -----------------------------------------------------------
// 4. M3U ÇIKTISINI OLUŞTURMA
// -----------------------------------------------------------

foreach ($streams as $stream) {
    // Gerekli bilgileri kontrol et
    if (empty($stream['stream_id']) || empty($stream['name'])) {
        continue;
    }

    $stream_id = $stream['stream_id'];
    $name = $stream['name'];
    $icon = isset($stream['stream_icon']) ? $stream['stream_icon'] : '';
    $category_id = isset($stream['category_id']) ? $stream['category_id'] : 'Diğer';
    
    // Kategori adını bul
    $category_name = isset($category_map[$category_id]) ? $category_map[$category_id] : 'Diğer';
    
    // Genişletilmiş Bilgi Satırı (#EXTINF)
    // xtream-codes formatında ek bilgiler eklenir
    echo "#EXTINF:-1 tvg-id=\"" . $stream_id . "\"";
    echo " tvg-name=\"" . $name . "\"";
    echo " tvg-logo=\"" . $icon . "\"";
    echo " group-title=\"" . $category_name . "\",";
    echo $name . "\n";

    // Akış URL'si (Xtream Codes API akış formatı)
    $container = isset($stream['container_extension']) ? $stream['container_extension'] : 'ts';
    $stream_url = $host . "/live/" . $username . "/" . $password . "/" . $stream_id . "." . $container;
    
    echo $stream_url . "\n";
}

?>
